File: src/Features/CSharp/Portable/CodeRefactorings/NodeSelectionHelpers.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

#nullable disable

using System;
using System.Collections.Immutable;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis.CodeRefactorings;
using Microsoft.CodeAnalysis.CSharp.LanguageService;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings
{
    internal static class NodeSelectionHelpers
    {
        internal static async Task<ImmutableArray<SyntaxNode>> GetSelectedDeclarationsOrVariablesAsync(CodeRefactoringContext context)
        {
            var (document, span, cancellationToken) = context;
            if (span.IsEmpty)
            {
                // if the span is empty then we are only selecting one "member" (which could include a field which declared multiple actual members)
                // Consider:
                // MemberDeclaration: member that can be declared in type (those are the ones we can pull up) 
                // VariableDeclaratorSyntax: for fields the MemberDeclaration can actually represent multiple declarations, e.g. `int a = 0, b = 1;`.
                // ..Since the user might want to select & pull up only one of them (e.g. `int a = 0, [|b = 1|];` we also look for closest VariableDeclaratorSyntax.
                var memberDeclaration = await context.TryGetRelevantNodeAsync<MemberDeclarationSyntax>().ConfigureAwait(false);
                if (memberDeclaration == null)
                {
                    // could not find a member, we may be directly on a variable declaration
                    var varDeclarator = await context.TryGetRelevantNodeAsync<VariableDeclaratorSyntax>().ConfigureAwait(false);
                    return varDeclarator == null
                        ? ImmutableArray<SyntaxNode>.Empty
                        : ImmutableArray.Create<SyntaxNode>(varDeclarator);
                }
                else
                {
                    return memberDeclaration switch
                    {
                        FieldDeclarationSyntax fieldDeclaration => fieldDeclaration.Declaration.Variables.AsImmutable<SyntaxNode>(),
                        EventFieldDeclarationSyntax eventFieldDeclaration => eventFieldDeclaration.Declaration.Variables.AsImmutable<SyntaxNode>(),
                        IncompleteMemberSyntax or GlobalStatementSyntax => ImmutableArray<SyntaxNode>.Empty,
                        _ => ImmutableArray.Create<SyntaxNode>(memberDeclaration),
                    };
                }
            }
            else
            {
                // if the span is non-empty, then we get potentially multiple members
                // Note: even though this method handles the empty span case, we don't use it because it doesn't correctly
                // pick up on keywords before the declaration, such as "public static int".
                // We could potentially use it for every case if that behavior changes
                var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
                var members = await CSharpSelectedMembers.Instance.GetSelectedMembersAsync(tree, span, allowPartialSelection: true, cancellationToken).ConfigureAwait(false);
                // if we get a node that would not have an obtainable symbol (such as the ones below)
                // we return an empty list instead of filtering so we don't get other potentially
                // malformed syntax nodes.
                // Consider pub[||] static int Foo;
                // Which has 2 member nodes (an incomplete and a field), but we'd only expect one
                return members.Any(m => m is GlobalStatementSyntax or IncompleteMemberSyntax)
                    ? ImmutableArray<SyntaxNode>.Empty
                    : members;
            }
        }
    }
}

File: src/Features/VisualBasic/Portable/CodeRefactorings/NodeSelectionHelpers.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis.CodeRefactorings
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.LanguageService
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings
    Friend Module NodeSelectionHelpers
        Friend Async Function GetSelectedMemberDeclarationAsync(context As CodeRefactoringContext) As Task(Of ImmutableArray(Of SyntaxNode))
            Dim document As Document = context.Document
            Dim span As TextSpan = context.Span
            Dim cancellationToken As CancellationToken = context.CancellationToken
            If span.IsEmpty Then
                ' MethodBaseSyntax also includes properties
                Dim methodMember = Await context.TryGetRelevantNodeAsync(Of MethodBaseSyntax)().ConfigureAwait(False)
                If methodMember IsNot Nothing Then
                    Return ImmutableArray.Create(Of SyntaxNode)(methodMember)
                End If
                ' Gets field variable declarations (not including the keywords like Public/Shared, etc), which are not methods
                Dim fieldDeclaration = Await context.TryGetRelevantNodeAsync(Of FieldDeclarationSyntax).ConfigureAwait(False)
                If fieldDeclaration Is Nothing Then
                    ' Gets the identifier + type of the field itself (ex. TestField As Integer), since it is nested in the variable declaration
                    ' And so the token's parent is not a variable declaration
                    Dim modifiedIdentifier = Await context.TryGetRelevantNodeAsync(Of ModifiedIdentifierSyntax).ConfigureAwait(False)
                    If modifiedIdentifier Is Nothing Then
                        Return ImmutableArray(Of SyntaxNode).Empty
                    Else
                        Return ImmutableArray.Create(Of SyntaxNode)(modifiedIdentifier)
                    End If
                Else
                    ' Field declarations can contain multiple variables (each of which are a "member")
                    Return fieldDeclaration.Declarators.SelectMany(Function(vds) vds.Names).Cast(Of SyntaxNode).AsImmutable()
                End If
            Else
                ' if the span is non-empty, then we get potentially multiple members
                ' Note: even though this method handles the empty span case, we don't use it because it doesn't correctly
                ' pick up on keywords before the declaration, such as "public static int".
                ' We could potentially use it for every case if that behavior changes
                Dim tree = Await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(False)
                Dim selectedMembers = Await VisualBasicSelectedMembers.Instance.
                    GetSelectedMembersAsync(tree, span, allowPartialSelection:=True, cancellationToken).ConfigureAwait(False)
                If selectedMembers.OfType(Of IncompleteMemberSyntax)().Any() Then
                    Return ImmutableArray(Of SyntaxNode).Empty
                Else
                    Return selectedMembers
                End If
            End If
        End Function
    End Module
End Namespace

